File: app/src/main/java/it/unical/mat/embasp/base/AnswerSet.java
package it.unical.mat.embasp.base;

import java.lang.reflect.InvocationTargetException;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import it.unical.mat.embasp.mapper.ASPMapper;

/**
 * <p>AnswerSet class rapresents an Answer Set and contains get and set methods to handle it.</p>
 * @see java.util.HashMap
 */
public class AnswerSet {
    private List<String> answerSet;//String representing the Answer Set
    private HashMap<Integer,Integer> weightMap;//Answer sets weights
    private Set<Object> objectsAtoms;

    /**
     * Constructor intialize an AnswerSet object with a String containing the Answer Set
     * and an {@link java.util.HashMap} containing Answer Set weight
     * @param outputString String representing an Answer Set
     */
    public AnswerSet(List<String> outputString){
        this.answerSet = outputString;
        this.weightMap = new HashMap<Integer, Integer>();
    }

    /**
     * Get function for the Answer Set in String format
     * @return answerSet Answer Set in String format
     */
    public List<String> getAnswerSet(){
        return Collections.unmodifiableList(this.answerSet);
    }

    public Set<Object> getAnswerObjects() throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
        if(objectsAtoms ==null) {
            objectsAtoms = new HashSet<>();
            ASPMapper mapper = ASPMapper.getInstance();
            for (String atom : answerSet) {
                    Object obj=mapper.getObject(atom);
                    if(obj!=null)
                        objectsAtoms.add(obj);
            }
        }
        return objectsAtoms;
    }

}

File: CSharp version/it/unical/mat/embasp/languages/asp/AnswerSet.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace it.unical.mat.embasp.languages.asp
{
    public class AnswerSet
    {
        private readonly IList<string> value;
        private readonly IDictionary<int, int> weight_map;
        private ISet<object> atoms;

        public AnswerSet(IList<string> output)
        {
            value = output;
            weight_map = new Dictionary<int, int>();
        }

        public AnswerSet(IList<string> value, IDictionary<int, int> weightMap)
        {
            this.value = value;
            weight_map = weightMap;
        }

        public virtual IList<string> GetAnswerSet() => new ReadOnlyCollection<string>(value);

        public virtual IList<string> Value => value;

        public virtual ISet<object> Atoms
        {
            get
            {
                if (atoms == null)
                {
                    atoms = new HashSet<object>();
                    ASPMapper mapper = ASPMapper.Instance;
                    foreach (String atom in value)
                    {
                        object obj = mapper.GetObject(atom);
                        if (obj != null)
                            atoms.Add(obj);
                    }
                }
                return atoms;
            }
        }

        public virtual IDictionary<int, int> LevelWeight => weight_map;

        public virtual IDictionary<int, int> Weights => new ReadOnlyDictionary<int, int>(weight_map);

        public override string ToString() => value.ToString();
    }
}